File: chat-sdk/src/main/java/ru/usedesk/chat_sdk/internal/di/MainModule.java
package ru.usedesk.chat_sdk.internal.di;

import android.content.ContentResolver;
import android.content.Context;

import com.google.gson.Gson;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.schedulers.Schedulers;
import ru.usedesk.chat_sdk.external.IUsedeskChat;
import ru.usedesk.chat_sdk.external.entity.IUsedeskActionListener;
import ru.usedesk.chat_sdk.external.entity.UsedeskActionListenerRx;
import ru.usedesk.chat_sdk.external.entity.UsedeskChatConfiguration;
import ru.usedesk.chat_sdk.external.service.notifications.presenter.UsedeskNotificationsPresenter;
import ru.usedesk.chat_sdk.internal.data.framework.api.apifile.FileApi;
import ru.usedesk.chat_sdk.internal.data.framework.api.apifile.IFileApi;
import ru.usedesk.chat_sdk.internal.data.framework.configuration.ConfigurationLoader;
import ru.usedesk.chat_sdk.internal.data.framework.file.FileLoader;
import ru.usedesk.chat_sdk.internal.data.framework.file.IFileLoader;
import ru.usedesk.chat_sdk.internal.data.framework.fileinfo.FileInfoLoader;
import ru.usedesk.chat_sdk.internal.data.framework.fileinfo.IFileInfoLoader;
import ru.usedesk.chat_sdk.internal.data.framework.httpapi.HttpApiLoader;
import ru.usedesk.chat_sdk.internal.data.framework.httpapi.IHttpApiLoader;
import ru.usedesk.chat_sdk.internal.data.framework.info.DataLoader;
import ru.usedesk.chat_sdk.internal.data.framework.loader.TokenLoader;
import ru.usedesk.chat_sdk.internal.data.framework.multipart.IMultipartConverter;
import ru.usedesk.chat_sdk.internal.data.framework.multipart.MultipartConverter;
import ru.usedesk.chat_sdk.internal.data.framework.socket.SocketApi;
import ru.usedesk.chat_sdk.internal.data.repository.api.ApiRepository;
import ru.usedesk.chat_sdk.internal.data.repository.api.IApiRepository;
import ru.usedesk.chat_sdk.internal.data.repository.configuration.IUserInfoRepository;
import ru.usedesk.chat_sdk.internal.data.repository.configuration.UserInfoRepository;
import ru.usedesk.chat_sdk.internal.domain.ChatInteractor;
import ru.usedesk.common_sdk.internal.api.IUsedeskApiFactory;
import ru.usedesk.common_sdk.internal.api.UsedeskApiFactory;
import ru.usedesk.common_sdk.internal.api.UsedeskOkHttpClientFactory;
import toothpick.config.Module;

class MainModule extends Module {

    MainModule(@NonNull Context appContext,
               @NonNull UsedeskChatConfiguration usedeskChatConfiguration,
               @NonNull IUsedeskActionListener actionListener) {
        bind(Context.class).toInstance(appContext);
        bind(ContentResolver.class).toInstance(appContext.getContentResolver());
        bind(UsedeskChatConfiguration.class).toInstance(usedeskChatConfiguration);
        bind(IUsedeskActionListener.class).toInstance(actionListener);

        bind(IUsedeskChat.class).to(ChatInteractor.class).singleton();

        bind(IUserInfoRepository.class).to(UserInfoRepository.class).singleton();
        bind(IApiRepository.class).to(ApiRepository.class).singleton();

        bind(DataLoader.class).withName("configuration").to(ConfigurationLoader.class).singleton();
        bind(DataLoader.class).withName("token").to(TokenLoader.class).singleton();

        bind(IMultipartConverter.class).to(MultipartConverter.class).singleton();
        bind(IFileApi.class).to(FileApi.class).singleton();
        bind(IUsedeskApiFactory.class).to(UsedeskApiFactory.class).singleton();
        bind(IFileLoader.class).to(FileLoader.class).singleton();
        bind(SocketApi.class).to(SocketApi.class).singleton();
        bind(IFileInfoLoader.class).to(FileInfoLoader.class).singleton();
        bind(IHttpApiLoader.class).to(HttpApiLoader.class).singleton();

        bind(UsedeskOkHttpClientFactory.class).to(UsedeskOkHttpClientFactory.class).singleton();
        bind(Gson.class).toInstance(gson());

        bind(Scheduler.class).withName("work").toInstance(Schedulers.io());
        bind(Scheduler.class).withName("main").toInstance(AndroidSchedulers.mainThread());

        //tmp for service
        bind(UsedeskNotificationsPresenter.class).to(UsedeskNotificationsPresenter.class).singleton();
        bind(UsedeskActionListenerRx.class).to(UsedeskActionListenerRx.class).singleton();
    }

    private Gson gson() {
        return new Gson();
    }
}

File: chat-sdk/src/main/java/ru/usedesk/chat_sdk/internal/di/MainModule.kt
package ru.usedesk.chat_sdk.internal.di

import android.content.Context
import com.google.gson.Gson
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ru.usedesk.chat_sdk.external.IUsedeskChat
import ru.usedesk.chat_sdk.external.entity.IUsedeskActionListener
import ru.usedesk.chat_sdk.external.entity.UsedeskChatConfiguration
import ru.usedesk.chat_sdk.external.service.notifications.presenter.UsedeskNotificationsPresenter
import ru.usedesk.chat_sdk.internal.data.framework._extra.retrofit.HttpApiFactory
import ru.usedesk.chat_sdk.internal.data.framework._extra.retrofit.IHttpApiFactory
import ru.usedesk.chat_sdk.internal.data.framework.configuration.ConfigurationLoader
import ru.usedesk.chat_sdk.internal.data.framework.configuration.IConfigurationLoader
import ru.usedesk.chat_sdk.internal.data.framework.fileinfo.FileInfoLoader
import ru.usedesk.chat_sdk.internal.data.framework.fileinfo.IFileInfoLoader
import ru.usedesk.chat_sdk.internal.data.framework.httpapi.HttpApiLoader
import ru.usedesk.chat_sdk.internal.data.framework.httpapi.IHttpApiLoader
import ru.usedesk.chat_sdk.internal.data.framework.token.ITokenLoader
import ru.usedesk.chat_sdk.internal.data.framework.token.TokenLoader
import ru.usedesk.chat_sdk.internal.data.repository.api.ApiRepository
import ru.usedesk.chat_sdk.internal.data.repository.api.IApiRepository
import ru.usedesk.chat_sdk.internal.data.repository.api.loader.SocketApi
import ru.usedesk.chat_sdk.internal.data.repository.configuration.IUserInfoRepository
import ru.usedesk.chat_sdk.internal.data.repository.configuration.UserInfoRepository
import ru.usedesk.chat_sdk.internal.domain.ChatInteractor
import toothpick.config.Module

internal class MainModule(
        appContext: Context,
        usedeskChatConfiguration: UsedeskChatConfiguration,
        actionListener: IUsedeskActionListener
) : Module() {

    init {
        bind(Context::class.java).toInstance(appContext)
        bind(UsedeskChatConfiguration::class.java).toInstance(usedeskChatConfiguration)
        bind(IUsedeskActionListener::class.java).toInstance(actionListener)
        bind(Scheduler::class.java).withName("work").toInstance(Schedulers.io())
        bind(Scheduler::class.java).withName("main").toInstance(AndroidSchedulers.mainThread())
        bind(Gson::class.java).toInstance(gson())

        bind(SocketApi::class.java).to(SocketApi::class.java).singleton()
        bind(IConfigurationLoader::class.java).to(ConfigurationLoader::class.java).singleton()
        bind(ITokenLoader::class.java).to(TokenLoader::class.java).singleton()
        bind(IFileInfoLoader::class.java).to(FileInfoLoader::class.java).singleton()
        bind(IHttpApiLoader::class.java).to(HttpApiLoader::class.java).singleton()

        bind(IUserInfoRepository::class.java).to(UserInfoRepository::class.java).singleton()
        bind(IApiRepository::class.java).to(ApiRepository::class.java).singleton()
        bind(IHttpApiFactory::class.java).to(HttpApiFactory::class.java).singleton()

        bind(IUsedeskChat::class.java).to(ChatInteractor::class.java).singleton()

        //tmp for service
        bind(UsedeskNotificationsPresenter::class.java).to(UsedeskNotificationsPresenter::class.java).singleton()
    }

    private fun gson(): Gson {
        return Gson()
    }
}